apps/api/migrations/preference-centralization/preference-centralization-migration.ts

```diff
@@ -0,0 +1,202 @@
+    setTimeout(resolve, randomizedMs);
+  });
+};
+
+const counter: Record<string, { success: number; error: number }> = {
+  subscriberGlobal: { success: 0, error: 0 },
+  subscriberWorkflow: { success: 0, error: 0 },
+  subscriberUnknown: { success: 0, error: 0 },
+  workflow: { success: 0, error: 0 },
+};
+
+let lastProcessedWorkflowId: string | undefined;
+let lastProcessedSubscriberId: string | undefined;
+
+process.on('SIGINT', () => {
+  console.log('\nGracefully shutting down from SIGINT (Ctrl+C)');
+  console.log({ counter });
+  if (lastProcessedWorkflowId) {
+    console.log(`Last processed workflow preference ID: ${lastProcessedWorkflowId}`);
+  }
+  if (lastProcessedSubscriberId) {
+    console.log(`Last processed subscriber preference ID: ${lastProcessedSubscriberId}`);
+  }
+  process.exit();
+});
+
+/**
+ * Migration to centralize workflow and subscriber preferences.
+ * - subscriber global preference
+ *    -> preferences with subscriber global type
+ * - subscriber workflow preferences
+ *    -> preferences with subscriber workflow type
+ * - workflow preferences
+ *   -> preferences with workflow-resource type
+ *   -> preferences with user-workflow type
+ */
+export async function preferenceCentralization(startWorkflowId?: string, startSubscriberId?: string) {
+  const app = await NestFactory.create(AppModule, {
+    logger: false,
+  });
+  console.log('start migration - preference centralization');
+
+  const upsertPreferences = app.get(UpsertPreferences);
+  const subscriberPreferenceRepository = app.get(SubscriberPreferenceRepository);
+  const workflowPreferenceRepository = app.get(NotificationTemplateRepository);
+
+  await migrateWorkflowPreferences(workflowPreferenceRepository, upsertPreferences, startWorkflowId);
+  console.log({ counter });
+  await migrateSubscriberPreferences(subscriberPreferenceRepository, upsertPreferences, startSubscriberId);
+
+  console.log('end migration - preference centralization');
+  console.log({ counter });
+  console.log(`Processed workflow preference with ID: ${lastProcessedWorkflowId}`);
+  console.log(`Processed subscriber preference with ID: ${lastProcessedSubscriberId}`);
+
+  app.close();
+}
+
+async function migrateWorkflowPreferences(
+  workflowPreferenceRepository: NotificationTemplateRepository,
+  upsertPreferences: UpsertPreferences,
+  startWorkflowId?: string
+) {
+  console.log('start workflow preference migration');
+  let query = {};
+  if (startWorkflowId) {
+    console.log(`Starting from workflow preference ID: ${startWorkflowId}`);
+    query = { _id: { $gt: startWorkflowId } };
+  }
+  const workflowPreferenceCursor = await workflowPreferenceRepository._model
+    .find(query)
+    .sort({ _id: 1 })
+    .batchSize(BATCH_SIZE)
+    .cursor();
+
+  for await (const workflowPreference of workflowPreferenceCursor) {
+    await sleep(10);
+    try {
+      const userWorkflowPreferenceToUpsert = UpsertUserWorkflowPreferencesCommand.create({
```

Comment:
Do we need to migrate user workflow preferences before workflow code-specific preferences?

Reply:
I added a transaction around both to alleviate the concern entirely.